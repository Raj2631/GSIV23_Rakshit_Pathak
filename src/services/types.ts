export interface Movie {
  adult: boolean;
  backdrop_path: string;
  genre_ids?: number[] | null;
  id: number;
  original_language: string;
  original_title: string;
  overview: string;
  popularity: number;
  poster_path: string;
  release_date: string;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
  runtime?: number;
}
export interface MoviesResponse {
  page: number;
  total_pages: number;
  total_results: number;
  results: Movies;
}

export interface Cast {
  adult: boolean;
  gender: number;
  id: number;
  known_for_department: string;
  name: string;
  original_name: string;
  popularity: number;
  profile_path: string;
  cast_id: number;
  character: string;
  credit_id: string;
  order: number;
}

export interface Crew {
  adult: boolean;
  gender: number;
  id: number;
  known_for_department: string;
  name: string;
  original_name: string;
  popularity: number;
  profile_path: string;
  credit_id: string;
  department: string;
  job: string;
}

export interface MovieCreditsResponse {
  id: number;
  crew: Crew[];
  cast: Cast[];
}

export interface CastsAndDirector {
  cast: Cast[];
  director: string;
}
export type Movies = Movie[];
