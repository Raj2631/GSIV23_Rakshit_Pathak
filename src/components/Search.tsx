import { SearchIcon } from "lucide-react";
import React, { useState } from "react";

const Search = () => {
  return (
    <div className="relative flex-grow max-w-sm bg-gray-300  rounded-xl">
      <form
        onSubmit={(e) => {
          e.preventDefault();
        }}
        className=" flex items-center bg-darkPurple rounded-full"
      >
        <SearchIcon size={20} className="ml-2 text-gray-500" strokeWidth={3} />
        <input
          type="text"
          placeholder="Search"
          className="p-1.5 border-none outline-none bg-transparent font-semibold flex-grow"
        />
      </form>
    </div>
  );
};

export default Search;
