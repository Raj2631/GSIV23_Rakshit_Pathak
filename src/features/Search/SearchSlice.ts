import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

export interface SearchInputState {
  value: string;
  isTyping: boolean;
}

const initialState: SearchInputState = {
  value: "",
  isTyping: false,
};

export const SearchSlice = createSlice({
  name: "searchInput",
  initialState,
  reducers: {
    setSearchValue: (state, action: PayloadAction<string>) => {
      state.value = action.payload;
      state.isTyping = false;
    },

    setIsTyping: (state, action: PayloadAction<boolean>) => {
      state.isTyping = action.payload;
    },
  },
});

export const { setSearchValue, setIsTyping } = SearchSlice.actions;

export default SearchSlice.reducer;
