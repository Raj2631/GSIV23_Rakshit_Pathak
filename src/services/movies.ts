import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { Movie, PopularMoviesResponse } from "./types";

export const moviesApi = createApi({
  reducerPath: "moviesApi",
  baseQuery: fetchBaseQuery({
    baseUrl: import.meta.env.VITE_MOVIEDB_API,
    prepareHeaders(headers) {
      headers.set(
        "Authorization",
        `Bearer ${import.meta.env.VITE_MOVIEDB_TOKEN}`
      );
    },
  }),
  endpoints: (builder) => ({
    getPopularMovies: builder.query<PopularMoviesResponse, number>({
      query: (page = 1) => `/movie/popular?language=en-US&page=${page}`,
      serializeQueryArgs: ({ endpointName }) => {
        return endpointName;
      },
      merge: (currentCache, newItems) => {
        currentCache.page = newItems.page;
        currentCache.results.push(...newItems.results);
      },
      forceRefetch({ currentArg, previousArg }) {
        return currentArg !== previousArg;
      },
    }),
    getMovieById: builder.query<Movie, string>({
      query: (movieId) => `/movie/${movieId}?language=en-US`,
    }),
  }),
});

export const { useGetPopularMoviesQuery, useGetMovieByIdQuery } = moviesApi;
