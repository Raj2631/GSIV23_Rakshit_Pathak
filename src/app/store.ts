import { configureStore } from "@reduxjs/toolkit";
import { moviesApi } from "../services/movies";
import SearchReducer from "../features/Search/SearchSlice";

export const store = configureStore({
  reducer: {
    [moviesApi.reducerPath]: moviesApi.reducer,
    searchInput: SearchReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(moviesApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
