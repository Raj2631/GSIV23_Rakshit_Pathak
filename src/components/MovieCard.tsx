import { Link } from "react-router-dom";
import { Movie } from "../services/types";
import { ImgBaseURL } from "../utils/constants";

type Props = {
  movie: Movie;
};

const MovieCard = ({ movie }: Props) => {
  const { title, overview, vote_average } = movie;
  return (
    <div className="w-52 rounded-lg cursor-pointer overflow-hidden shadow-[rgba(50,50,93,0.25)_0px_6px_12px_-2px,_rgba(0,0,0,0.3)_0px_3px_7px_-3px]">
      <Link to={`/movieDetail/${movie.id}`}>
        <img
          key={movie.id}
          src={`${ImgBaseURL}/${movie.poster_path}`}
          className="w-full bg-black bg-cover"
        />
        <div className=" py-2 px-2">
          <div className="flex items-center  justify-between text-sm">
            <div className="w-24 truncate text-left font-semibold">{title}</div>
            <div className="text-gray-400">({vote_average})</div>
          </div>
          <div className="w-full text-sm text-left my-1 text-gray-500 line-clamp-2">
            {overview}
          </div>
        </div>
      </Link>
    </div>
  );
};

export default MovieCard;
