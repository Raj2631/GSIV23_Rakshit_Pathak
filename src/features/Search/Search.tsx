import { ChangeEvent } from "react";
import { SearchIcon, XIcon } from "lucide-react";
import { setSearchValue } from "./SearchSlice";
import { useAppDispatch, useAppSelector } from "../../app/hooks";

const Search = () => {
  const dispatch = useAppDispatch();
  const searchValue = useAppSelector((state) => state.searchInput.value);

  const onInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    dispatch(setSearchValue(e.target.value));
  };

  const clearInput = () => {
    dispatch(setSearchValue(""));
  };

  return (
    <div className="relative flex-grow max-w-sm bg-gray-300  rounded-xl">
      <form
        onSubmit={(e) => {
          e.preventDefault();
        }}
        className=" flex items-center bg-darkPurple rounded-full"
      >
        <SearchIcon size={20} className="ml-2 text-gray-500" strokeWidth={3} />
        <input
          type="text"
          placeholder="Search"
          className="p-1.5 border-none outline-none bg-transparent font-semibold flex-grow"
          value={searchValue}
          onChange={onInputChange}
        />
        {searchValue && (
          <button onClick={clearInput}>
            <XIcon size="18" className=" text-gray-500 mx-4" />
          </button>
        )}
      </form>
    </div>
  );
};

export default Search;
