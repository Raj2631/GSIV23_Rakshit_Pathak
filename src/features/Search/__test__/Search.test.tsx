import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Search from "../Search";
import { Provider } from "react-redux";
import { store } from "../../../app/store";
import { moviesApi } from "../../../services/movies";
import Loading from "../../../components/Loading";
import MovieCard from "../../../components/MovieCard";
import { BrowserRouter as Router } from "react-router-dom";

describe("Search Input component", () => {
  it("renders correctly", () => {
    render(
      <Provider store={store}>
        <Search />
      </Provider>
    );
    const searchInput = screen.getByPlaceholderText("Search");
    expect(searchInput).toBeInTheDocument();
  });

  it("Renders search result when input changes", async () => {
    render(
      <Provider store={store}>
        <Search />
      </Provider>
    );
    const searchInput = screen.getByPlaceholderText("Search");

    fireEvent.change(searchInput, { target: { value: "test" } });

    await waitFor(() =>
      expect(store.getState().searchInput.value).toEqual("test")
    );

    function Movies() {
      const { data, error, isLoading } =
        moviesApi.endpoints.getMoviesByTitle.useQuery({
          title: store.getState().searchInput.value,
          page: 1,
        });

      if (isLoading) {
        return <Loading />;
      }

      if (error) {
        <p>Error</p>;
      }

      const movie = data?.results[0];
      return <div>{movie && <MovieCard key={movie.id} movie={movie} />}</div>;
    }

    const tree = render(
      <Provider store={store}>
        <Router>
          <Movies />
        </Router>
      </Provider>
    );

    expect(screen.getByText(/Loading.../i)).toBeInTheDocument();

    await waitFor(async () => {
      expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
    });

    expect(await screen.findByText("Test")).toBeInTheDocument();
    expect(tree).toMatchSnapshot();
  });
});
