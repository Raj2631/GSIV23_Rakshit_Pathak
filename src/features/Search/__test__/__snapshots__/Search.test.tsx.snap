// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search Input component > Renders search result when input changes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex-grow max-w-sm bg-gray-300  rounded-xl"
      >
        <form
          class=" flex items-center bg-darkPurple rounded-full"
        >
          <svg
            class="ml-2 text-gray-500"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <path
              d="m21 21-4.3-4.3"
            />
          </svg>
          <input
            class="p-1.5 border-none outline-none bg-transparent font-semibold flex-grow"
            placeholder="Search"
            type="text"
            value="test"
          />
          <button>
            <svg
              class=" text-gray-500 mx-4"
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 6 6 18"
              />
              <path
                d="m6 6 12 12"
              />
            </svg>
          </button>
        </form>
      </div>
    </div>
    <div>
      <div>
        <div
          class="w-60 h-[444px] rounded-lg cursor-pointer overflow-hidden shadow-[rgba(50,50,93,0.25)_0px_6px_12px_-2px,_rgba(0,0,0,0.3)_0px_3px_7px_-3px]"
        >
          <a
            href="/movieDetail/226979"
          >
            <img
              alt="Test"
              class="w-full h-[360px] bg-black bg-cover"
              src="https://image.tmdb.org/t/p/w500//tTWRomgIMOoIB3CJLPlVbqSawEm.jpg"
            />
            <div
              class=" py-2 px-2"
            >
              <div
                class="flex items-center  justify-between text-sm"
              >
                <div
                  class="w-24 truncate text-left font-semibold"
                >
                  Test
                </div>
                <div
                  class="text-gray-400"
                >
                  (
                  6.685
                  )
                </div>
              </div>
              <div
                class="w-full text-sm text-left my-1 text-gray-500 line-clamp-2"
              >
                San Francisco, 1985. Two opposites attract at a modern dance company. Together, their courage and resilience are tested as they navigate a world full of risks and promise, against the backdrop of a disease no one seems to know anything about.
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="w-60 h-[444px] rounded-lg cursor-pointer overflow-hidden shadow-[rgba(50,50,93,0.25)_0px_6px_12px_-2px,_rgba(0,0,0,0.3)_0px_3px_7px_-3px]"
      >
        <a
          href="/movieDetail/226979"
        >
          <img
            alt="Test"
            class="w-full h-[360px] bg-black bg-cover"
            src="https://image.tmdb.org/t/p/w500//tTWRomgIMOoIB3CJLPlVbqSawEm.jpg"
          />
          <div
            class=" py-2 px-2"
          >
            <div
              class="flex items-center  justify-between text-sm"
            >
              <div
                class="w-24 truncate text-left font-semibold"
              >
                Test
              </div>
              <div
                class="text-gray-400"
              >
                (
                6.685
                )
              </div>
            </div>
            <div
              class="w-full text-sm text-left my-1 text-gray-500 line-clamp-2"
            >
              San Francisco, 1985. Two opposites attract at a modern dance company. Together, their courage and resilience are tested as they navigate a world full of risks and promise, against the backdrop of a disease no one seems to know anything about.
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
